JotunnLib/Managers/DungeonManager.cs

```diff
@@ -0,0 +1,372 @@
+                        }
+                    }
+                    catch (MockResolveException ex)
+                    {
+                        Logger.LogWarning(string.Format("Skipping Room {0}: could not resolve mock {1} {2}", customRoom, ex.MockType.Name, ex.FailedMockName));
+                        toDelete.Add(customRoom.Name);
+                    }
+                    catch (Exception ex2)
+                    {
+                        Logger.LogWarning(string.Format("Exception caught while adding Room: {0}", ex2));
+                        toDelete.Add(customRoom.Name);
+                    }
+                }
+                foreach (string name in toDelete)
+                {
+                    Rooms.Remove(name);
+                }
+
+                DungeonDB.instance.GenerateHashList();
+                GenerateHashes();
+            }
+        }
+
+        private void OnDungeonGeneratorSetupAvailableRooms(DungeonGenerator self)
+        {
+            DungeonGeneratorTheme proxy = self?.gameObject.GetComponent<DungeonGeneratorTheme>();
+
+            if (DungeonGenerator.m_availableRooms != null)
+            {
+                if (proxy != null)
+                {
+                    Logger.LogDebug($"This dungeon generator has a custom theme = {proxy.m_themeName}, adding available rooms");
+
+                    DungeonGenerator.m_availableRooms.AddRange(Rooms.Values
+                        .Where(r => r.Room.m_enabled)
+                        .Where(r => r.ThemeName == proxy.m_themeName)
+                        .Select(r => r.RoomData));
+                }
+                else
+                {
+                    Logger.LogDebug($"Adding additional rooms of type {self.m_themes} to available rooms");
+
+                    DungeonGenerator.m_availableRooms.AddRange(Rooms.Values
+                        .Where(r => r.Room.m_enabled)
+                        .Where(r => self.m_themes.HasFlag(r.Theme))
+                        .Select(r => r.RoomData));
+                }
+
+                if (DungeonGenerator.m_availableRooms.Count <= 0)
+                {
+                    Logger.LogDebug($"DungeonManager's SetupAvailableRooms yielded zero rooms.");
+                }
+            }
+        }
+
+
+        private void InvokeOnVanillaRoomsAvailable()
+        {
+            OnVanillaRoomsAvailable?.SafeInvoke();
+            OnRoomsRegistered?.SafeInvoke();
+        }
+
+        private void RegisterRoomInDungeonDB(CustomRoom room)
+        {
+            DungeonDB.instance.m_rooms.Add(room.RoomData);
+        }
+
+        /// <summary>
+        ///     Attempt to get room by hash.
+        /// </summary>
+        /// <param name="hash"></param>
+
+        private DungeonDB.RoomData OnDungeonDBGetRoom(int hash)
+        {
+            if (hashToName.TryGetValue(hash, out var roomName))
+            {
+                CustomRoom room = Rooms[roomName];
```

Comment:
accessing `Rooms[roomName]` without sanity checks looks a bit dangerous because of a possible KeyNotFoundException

Reply:
Addressed in https://github.com/Valheim-Modding/Jotunn/pull/434/commits/2d3bb465b3a467152e8f9e8d137fbed2cedb4d1c